### ISSUE TITLE
Inscriber fails when fed items directly from a Redstone Furnace

### ISSUE THREAD

**Issue by Lodran** [AppliedEnergistics/Applied-Energistics-2#1563]
I set up the first stage of a processor fab as follows:

An export bus feeds Certus Quarts Dust to a Redstone Furnace (Thermal Expansion).

The Redstone Furnace feeds silicon directly into an inscriber, containing a Silicon press.

The problem is that the inscriber finishes one Printed Silicon, and then never completes another - the charge indicator reaches 30% or so, and then drops backs to zero, repeatedly.

My guess is that when the Redstone furnace completes a silicon, and attempts to insert it into the inscriber, that it somehow causes the inscriber to reset its charge.

I was able to make the machine function by inserting an Extra Utilities transfer node between the Redstone Furnace and the Inscriber.

**Comment by yueh** [comment]
Yes, inscribers reset their progress when something changes the inventory or it becomes an invalid recipe (input only).

But it sounds more like TE itself is causing it as XU and other mod do not seem to cause it. Did you also report it to them?
Maybe they have some issues with handling the limited stacksize?

**Comment by KingLemming** [comment]
I know what's going on here and I'm not really sure where blame should go. We're both just being very efficient about things.

You guys are assuming that if a certain method is called, you should reset progress. That's...maybe sort of wrong. TE machines have similar functionality, but they check to see if the stack actually *has* changed. If not, progress isn't reset or affected in any way.

I'll ponder it, though I'm inclined to argue it's a potential design oversight on the part of the inscriber.

**Comment by yueh** [comment]
There are probably a few things on both sides to consider.

We basically assume nothing will mutate the inventory as it is full. If we consider a litmus test with vanilla, the one related to this is probably "do hoppers modify the inventory of a furnace when it is full".
And "full" for inscribers is really a binary condition as the stacksize is limited to one.

On the other hand we should probably not reset it, if both stacks are equal.
Even if it is just the binary condition on our side.

I personally would just blame `IInventory` for being badly designed and unnecessarily breaking encapsulation, thus exposing the internal inventory with `setInventorySlotContents`. Instead it should just be a simple `addToSlot` with the leftover as return. So it would allow us to actively prevent something inserting a full stack because it did ignore the inventory limitations. Instead of dumping them into the void, which would the only thing being currently possible.
And no longer being in limbo because of some undefined behaviour where every mod needs to consider the assumption of every other mod.

**Comment by KingLemming** [comment]
Yeah, IInventory leaves a *lot* to be desired. We use the same generic item insertion code for everything, so the question in my mind is the performance penalty involved in the extra check. And it's possible that maybe it's worth our while to have the check on our end. It kludges up the code slightly but would prevent stuff like this.

Hard to say really, the interface as designed isn't very specific. I pushed for an IInventoryHandler class a while ago, but the powers that be literally didn't understand it. They just don't have any grasp of how the game should really work.

**Comment by Blue64** [comment]
@KingLemming That made me lol, I completely understand where you're coming from, I've been trying to figure out a way to access the LiquidDict via MineTweaker/ModTweaker for days, I only want to do 2 simple things, but short of editing and recompiling the entire mod, I can't seem to figure out how to access the ILiquidDefinition.class to use the addContainer function, because PneumaticCraft derped and forgot to add their Oil Bucket to the liquid:oil registry (along with BuildCraft forcing priority for their Oil even when used on PneumaticCraft Oil TileEntities, it's all one huge dense highly flammable mess, but I digress), the other thing I'm trying to fix is the liquid:life essence LiquidDict entry, of all the mods I use, only 1 LiquidDict entry has a space -_-|| and it's one I wanted to add functionality for too (turn Mob Eggs into Life Essence, Silverfish Blocks too)
